packages/webdriverio/src/commands/browser/switchFrame.ts

```diff
@@ -0,0 +1,217 @@
+
+        /**
+         * transform context tree into a flat list of context objects with references
+         * to children
+         */
+        const allContexts: Record<string, FlatContextTree> = tree.contexts.map(mapContext).flat(Infinity)
+            .filter((ctx) => ctx !== currentContext)
+            .reduce((acc, ctx) => {
+                const context = findContextById(tree.contexts, ctx)
+                acc[ctx] = context
+                return acc
+            }, {} as Record<string, FlatContextTree>)
+
+        for (const [contextId, ctx] of Object.entries(allContexts)) {
+            sessionContext.setCurrentContext(contextId)
+            const isDesiredFrame = await context(ctx)
+            if (isDesiredFrame) {
+                return context
+            }
+        }
+
+        sessionContext.setCurrentContext(currentContext)
+        throw new Error('Could not find the desired frame')
+    }
+
+    throw new Error(
+        `Invalid type for context parameter: ${typeof context}, expected one of number, string or null. ` +
+        'Check out our docs: https://webdriver.io/docs/api/browser/switchToFrame.html'
+    )
+}
+
+function switchToFrameHelper (browser: WebdriverIO.Browser, context: string) {
+    const sessionContext = getContextManager(browser)
+    sessionContext.setCurrentContext(context)
+}
+
+async function switchToFrameUsingElement (browser: WebdriverIO.Browser, element: WebdriverIO.Element) {
+    let frameSrc = await element.getAttribute('src')
+    if (!frameSrc) {
+        throw new Error(`Frame with index ${context} has no src attribute`)
+    }
+
+    if (!frameSrc.startsWith('http')) {
+        /**
+         * use the browser to access `URL.parse` as it is not available in Node.js until v22
+         */
+        frameSrc = await browser.execute((urlPath: string) => (
+            URL.parse(urlPath, window.location.href)?.href
+        ), frameSrc) || frameSrc
+    }
+
+    const tree = await browser.browsingContextGetTree({})
+    const urlContext = findContext(frameSrc, tree.contexts, byUrl)
+    if (!urlContext) {
+        throw new Error(`Frame with url "${frameSrc}" not found`)
+    }
+
+    await switchToFrameHelper(browser, urlContext)
+    return urlContext
+}
+
+function byUrl (context: local.BrowsingContextInfo, url: string) {
+    return context.url === url
+}
+
+function byUrlContaining (context: local.BrowsingContextInfo, url: string) {
+    return context.url.includes(url)
+}
+
+function byContextId (context: local.BrowsingContextInfo, contextId: string) {
+    return context.context === contextId
+}
+
+function findContext (
+    url: string,
+    contexts: local.BrowsingContextInfoList,
+    matcher: typeof byUrl | typeof byUrlContaining
+): string | undefined {
+    for (const context of contexts) {
+        if (matcher(context, url)) {
+            return context.context
+        }
+
+        if (context.children && context.children.length > 0) {
```

Comment:
not sure if the ? needs to be added between context and children or between children and length or both but it's a bit cleaner
```suggestion
        if (context.children?.length > 0) {
```